### PR TITLE
Feat/import csv

### PR DESCRIPTION
**Pull request by Scttpr**
Checklist

- [x] I've added a mention in CHANGELOG.md

### CODE REVIEW

**Comment by Scttpr** [comment]
[Ajouter import CSV POI / Chauffeurs / Véhicules](https://trello.com/c/9O9BSm7K/133-ajouter-import-csv-poi-chauffeurs-v%C3%A9hicules)

File: middlewares/csv-to-json.mjs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  await next();
+};
+
+export const csvToJson = async (ctx, next) => {
+  const { length, 0: file } = Object.keys(ctx.request.files).map((key) => ctx.request.files[key]);
+  if (length > 1) {
+    ctx.throw_and_log(422, 'Please send one file at a time');
+  }
+
+  ctx.file = await csv2Json({ delimiter: ';', ignoreEmpty: true }).fromFile(file.path);
```
Comment (Nainterceptor) [inline middlewares/csv-to-json.mjs]: Is it possible to send delimiter and ignoreEmpty in the request ? Like in the csv generation of rides ? (history)

File: models/poi.mjs [hunk 2]
```diff
@@ -26,6 +28,7 @@ const PoiSchema = new Schema({
 });
 
 PoiSchema.plugin(createdAtPlugin);
+PoiSchema.plugin(addCSVContentPlugin, MODEL_NAME);
```
Comment (Nainterceptor) [inline models/poi.mjs]: MODEL_NAME is unused in the plugin

File: models/phone.mjs [hunk 3]
```diff
@@ -37,5 +39,6 @@ const PhoneSchema = new Schema({
 });
 
 PhoneSchema.plugin(createdAtPlugin);
+PhoneSchema.plugin(addCSVContentPlugin, MODEL_NAME);
```
Comment (Nainterceptor) [inline models/phone.mjs]: MODEL_NAME is unused in the plugin

File: models/phone-model.mjs [hunk 4]
```diff
@@ -1,14 +1,17 @@
...
 const { Schema } = mongoose;
+const MODEL_NAME = 'PhoneModel';
 
 const PhoneModelSchema = new Schema({
   _id: String,
   label: String,
 });
 
 PhoneModelSchema.plugin(createdAtPlugin);
+PhoneModelSchema.plugin(addCSVContentPlugin, MODEL_NAME);
```
Comment (Nainterceptor) [inline models/phone-model.mjs]: MODEL_NAME is unused in the plugin

File: models/helpers/add-csv-content.mjs [hunk 5]
```diff
@@ -0,0 +1,16 @@
+export default function addCSVContentPlugin(schema) {
+  // eslint-disable-next-line no-param-reassign
+  schema.statics.createFromCSV = async function createFromCSV(datas) {
+    const session = await this.startSession();
+    session.startTransaction();
+    try {
+      await this.insertMany(datas, { session });
```
Comment (Nainterceptor) [inline models/helpers/add-csv-content.mjs]: According to documentation, [insertMany](https://mongoosejs.com/docs/api.html#model_Model.insertMany) don't run `save` hook, and we're using this hook in some models :/

File: models/category.mjs [hunk 6]
```diff
@@ -1,13 +1,16 @@
...
 const { Schema } = mongoose;
+const MODEL_NAME = 'Category';
 
 const CategorySchema = new Schema({
   _id: String,
   label: { type: String, required: true },
 });
 
 CategorySchema.plugin(createdAtPlugin);
+CategorySchema.plugin(addCSVContentPlugin, MODEL_NAME);
```
Comment (Nainterceptor) [inline models/category.mjs]: MODEL_NAME is unused in the plugin

File: models/car.mjs [hunk 7]
```diff
@@ -19,6 +21,7 @@ const CarSchema = new Schema({
 });
 
 CarSchema.plugin(createdAtPlugin);
+CarSchema.plugin(addCSVContentPlugin, MODEL_NAME);
```
Comment (Nainterceptor) [inline models/car.mjs]: MODEL_NAME is unused in the plugin

File: models/car-model.mjs [hunk 8]
```diff
@@ -11,6 +13,8 @@ const CarModelSchema = new Schema({
 });
 
 CarModelSchema.plugin(createdAtPlugin);
+CarModelSchema.plugin(addCSVContentPlugin, MODEL_NAME);
```
Comment (Nainterceptor) [inline models/car-model.mjs]: MODEL_NAME is unused in the plugin

File: models/rating.mjs [hunk 9]
```diff
@@ -31,6 +33,7 @@ const RatingSchema = new Schema({
 });
 
 RatingSchema.plugin(createdAtPlugin);
+RatingSchema.plugin(addCSVContentPlugin, MODEL_NAME);
```
Comment (Nainterceptor) [inline models/rating.mjs]: MODEL_NAME is unused in the plugin

File: models/user.mjs [hunk 10]
```diff
@@ -116,6 +117,7 @@ const UserSchema = new Schema({
 
 UserSchema.plugin(createdAtPlugin);
 UserSchema.plugin(cleanObjectPlugin, MODEL_NAME);
+UserSchema.plugin(addCSVContentPlugin, MODEL_NAME);
```
Comment (Nainterceptor) [inline models/user.mjs]: MODEL_NAME is unused in the plugin

File: package.json [hunk 11]
```diff
@@ -67,6 +68,7 @@
     "winston-mongodb": "^5.0.1"
   },
   "devDependencies": {
+    "@babel/parser": "^7.7.5",
```
Comment (Nainterceptor) [inline package.json]: You can remove that, fix has been merged on develop

File: middlewares/csv-to-json.mjs [hunk 12]
```diff
@@ -0,0 +1,52 @@
+import csv2Json from 'csvtojson';
+import Campus from '../models/campus';
+
+export const checkDuplications = (Model, ref) => async (ctx, next) => {
+  const documents = await Model.find().lean();
+  const duplications = documents.filter((document) => {
+    // eslint-disable-next-line consistent-return
```
Comment (Nainterceptor) [inline middlewares/csv-to-json.mjs]: Here, eslint is right, don't disable this rule

File: middlewares/csv-to-json.mjs [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    ctx.file.forEach((data) => {
+      if (data[ref] === document[ref]) {
+        return true;
+      }
+    });
```
Comment (Nainterceptor) [inline middlewares/csv-to-json.mjs]: First, you can do the same thing in one-liner (which will be OK with consistent-return)
```javascript
ctx.file.forEach((data) => data[ref] === document[ref]);
```
But, ctx.file.forEach will return nothing. Please use reduce like (should be working but untested)
```javascript
const duplications = documents.filter((document) => ctx.file.reduce((acc, data) => (acc || data[ref] === document[ref]), false));
```

File: middlewares/csv-to-json.mjs [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+        ctx.throw_and_log(403, 'Campus does not match current one');
+      }
+      return newItem;
+    });
+  }
+  await next();
+};
+
+export const csvToJson = async (ctx, next) => {
+  const { body: { delimiter = ';', ignoreEmpty = true }, files } = ctx.request;
```
Comment (Nainterceptor) [inline middlewares/csv-to-json.mjs]: Can we put these parameters in query and not in body ? To be consistent with download (in history)

File: helpers/abstract-route.mjs [hunk 15]
```diff
@@ -3,6 +3,29 @@ import maskOutput from '../middlewares/mask-output';
 import resolveRights from '../middlewares/check-rights';
 import addFilter from '../middlewares/add-filter';
 
+export function addBatchToRouter(Model, {
+  url = '/batch', right, rights = [], middlewares = [], main, ref = '_id', refs = [],
```
Comment (Nainterceptor) [inline helpers/abstract-route.mjs]: ref & refs may be confusing

File: models/helpers/add-csv-content.mjs [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+        throw err;
+      }
+    });
+  }));
+};
+
+const checkDuplications = async ({
+  model: Model, ref, refs = [], datas,
+}) => {
+  refs.push(ref);
```
Comment (Nainterceptor) [inline models/helpers/add-csv-content.mjs]: Ok, so I think you should take only refs array, and in batch, you aggregate ref & refs, like for middlewares

File: models/helpers/add-csv-content.mjs [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+  // check duplications in batch
+  const batchDuplications = refs.reduce((acc, r) => {
+    const refData = datas.map((data) => data[r]);
+    refData.forEach((d) => {
+      let counter = 0;
+      refData.forEach((current) => {
```
Comment (Nainterceptor) [inline models/helpers/add-csv-content.mjs]: You can replace this forEach by a reduce.

Edit :
Like 
```javascript
const counter = refData.reduce((count, current) => current === d ? count+1 : count, 0);
```

File: models/helpers/add-csv-content.mjs [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+};
+
+const checkDuplications = async ({
+  model: Model, ref, refs = [], datas,
+}) => {
+  refs.push(ref);
+  // check duplications in batch
+  const batchDuplications = refs.reduce((acc, r) => {
+    const refData = datas.map((data) => data[r]);
+    refData.forEach((d) => {
```
Comment (Nainterceptor) [inline models/helpers/add-csv-content.mjs]: You can replace this forEach by a reduce too

File: models/helpers/add-csv-content.mjs [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+    const err = new Error();
+    err.status = 422;
+    err.message = `Duplications in the batch : ${batchDuplications}`;
+    throw err;
+  }
+
+  const documents = await Model.find().lean();
+  // check potential duplications in db
+  const dbDuplications = datas.reduce((acc, data) => {
+    refs.forEach((r) => {
```
Comment (Nainterceptor) [inline models/helpers/add-csv-content.mjs]: You can replace this forEach by a reduce too

File: models/helpers/add-csv-content.mjs [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+      let counter = 0;
+      documents.forEach((document) => {
```
Comment (Nainterceptor) [inline models/helpers/add-csv-content.mjs]: You can replace this forEach by a reduce too

**Pull request merged**